lib/awful/rules.lua

```diff
@@ -207,23 +211,169 @@ function rules.matches_list(c, _rules)
+
+    for _, v in ipairs(rules.rule_sources) do
+        -- Names must be unique
+        assert(
+            v.name ~= name,
+            "Name must be unique, but '" .. name .. "' was already registered."
+        )
+    end
+
+    rules._rule_source_sort:prepend(name, precede    )
+    rules._rule_source_sort:append (name, depends_on )
+
+    local res, err = rules._rule_source_sort:sort()
+
+    if err then
+        gdebug.print_warning("Failed to add the rule source: "..err)
+        return false
+    end
+
     local callbacks = {}
 
-    for _, entry in ipairs(rules.matching_rules(c, rules.rules)) do
-        if entry.properties then
-            for property, value in pairs(entry.properties) do
-                props[property] = value
-            end
+    -- Get all callbacks for *existing* sources.
+    -- It is important to remember that names can be used in the sorting even
+    -- if the source itself doesn't (yet) exists.
+    for _, v in ipairs(rules.rule_sources) do
+        callbacks[v.name] = v.callback
+    end
+
+    rules.rule_sources = {}
+    callbacks[name]    = callback
+
+    for _, v in ipairs(res) do
+        if callbacks[v] then
+            table.insert(rules.rule_sources, 1, {
+                callback = callbacks[v],
+                name     = v
+            })
+        end
+    end
+
+    return true
+end
+
+--- Remove a source.
+-- @tparam string name The source name.
+-- @treturn boolean If the source was removed
+function rules.remove_rule_source(name)
+    for k, v in ipairs(rules.rule_sources) do
+        if v.name == name then
+            table.remove(rules.rule_sources, k)
+            return true
         end
+    end
+
+    return false
+end
+
+-- Add the rules properties
+local function apply_awful_rules(c, props, callbacks)
+    for _, entry in ipairs(rules.matching_rules(c, rules.rules)) do
+        gtable.crush(props,entry.properties or {})
+
         if entry.callback then
             table.insert(callbacks, entry.callback)
         end
     end
+end
+
+--- The default `awful.rules` source.
+--
+-- **Depends on:**
+--
+-- *nothing*
+--
+-- @rulesources awful.rules
+
+rules.add_rule_source("awful.rules", apply_awful_rules, {"awful.spawn"}, {})
+
+-- Add startup_id overridden properties
+local function apply_spawn_rules(c, props, callbacks)
+    if c.startup_id and aspawn.snid_buffer[c.startup_id] then
+        local snprops, sncb = unpack(aspawn.snid_buffer[c.startup_id])
+
+        -- The SNID tag(s) always have precedence over the rules one(s)
+        if snprops.tag or snprops.tags or snprops.new_tag then
+            props.tag, props.tags, props.new_tag = nil, nil, nil
+        end
+
+        gtable.crush(props, snprops)
+        gtable.merge(callbacks, sncb)
+    end
+end
+
+--- The rule source for clients spawned by `awful.spawn`.
+--
+-- **Depends on:**
+--
+-- * `awful.rules`
+--
+-- @rulesources awful.spawn
+
+rules.add_rule_source("awful.spawn", apply_spawn_rules, {}, {"awful.rules"})
```

Comment:
Why is this in this file and not in `awful.spawn`?

Reply:
Because it creates an indirect circular module dependencies. `awful.rules` can require `awful.spawn`, but not the opposite.

Reply:
For future reference, the cycle is `spawn` -> `rules` -> `placement` -> `client` -> `spawn`.
Sigh. :-(